packages/edit-navigation/src/components/menu-locations-editor/index.js

```diff
@@ -1,3 +1,167 @@
+/**
+ * External dependencies
+ */
+import { map, flatMap, remove, find, pick, groupBy } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
+import {
+	Button,
+	Panel,
+	PanelBody,
+	SelectControl,
+	Spinner,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+const MenuSelectControl = ( { onChange, menusList, location } ) => {
+	const [ menuId, setMenuId ] = useState( false );
+
+	useEffect( () => {
+		setMenuId( location.menu );
+	}, [ location.menu ] );
+
+	return (
+		<SelectControl
+			options={ menusList }
+			value={ menuId }
+			onChange={ ( selectedMenuId ) => {
+				onChange( menuId, {
+					location: location.name,
+					menu: parseInt( selectedMenuId ),
+				} );
+				setMenuId( selectedMenuId );
+			} }
+		/>
+	);
+};
+
 export default function MenuLocationsEditor() {
-	return <>Menu locations editor</>;
+	const menus = useSelect( ( select ) => select( 'core' ).getMenus() );
+
+	const [ menuLocations, setMenuLocations ] = useState( false );
+	const [ locationsData, setLocationsData ] = useState( {} );
+	const { saveMenu } = useDispatch( 'core' );
+
+	const kind = 'root';
+	const name = 'menuLocation';
+	const entities = useSelect( ( select ) =>
+		select( 'core' ).getEntitiesByKind( kind )
+	);
+	const entity = find( entities, { kind, name } );
+
+	const getMenuLocations = async () => {
+		const path = `${ entity.baseURL }`;
+		let apiLocations = await apiFetch( {
+			path,
+			method: 'GET',
+		} );
+		apiLocations = flatMap( apiLocations, ( value ) => value );
+		return apiLocations;
+	};
+
+	const setLatestLocations = async () => {
+		const latestLocations = await getMenuLocations();
+		setMenuLocations( latestLocations );
+	};
+
+	useEffect( () => {
+		setLatestLocations();
+	}, [] );
+
+	useEffect( () => {
+		if ( menus && menuLocations ) {
+			const locations = groupBy( menuLocations, 'menu' );
+			map( locations, ( location, menuId ) => {
+				locations[ menuId ] = map( location, 'name' );
+			} );
+			setLocationsData( locations );
+		}
+	}, [ menuLocations ] );
+
+	if ( ! menus || ! menuLocations ) {
+		return <Spinner />;
+	}
+
+	const setLocations = ( prevMenuId, { location, menu } ) => {
+		remove( locationsData[ prevMenuId ], ( oldLocation ) => {
+			return oldLocation === location;
+		} );
+		if ( locationsData[ menu ] ) {
+			locationsData[ menu ].push( location );
+		} else {
+			locationsData[ menu ] = [ location ];
+		}
+		setLocationsData( locationsData );
+	};
+
+	const saveLocations = async () => {
+		for ( const menuId in locationsData ) {
+			if ( locationsData.hasOwnProperty( menuId ) ) {
+				const intMenuId = parseInt( menuId );
+				if ( intMenuId ) {
+					await saveMenu( {
+						...pick( find( menus, { id: intMenuId } ), [
```

Comment:
Name doesn't need to passed here. It isn't being updated. If a field is not passed, it will use the last value. So this pick/find shouldn't be required. Just the menuId.

Reply:
Hmm I updated this and its ok. For some reason at some point the API got me some error because no name was sent. Apparently it's ok now.